File: MinimumPathFallingSum.java
// TC:O(N*N) -> N*N is the size of matrix A
// SC:O(1) -> No extra space is required.
// Did it run successfully on LeetCode? Yes
class Solution {
    public int minFallingPathSum(int[][] A) {
        if (A == null || A.length == 0) 
            return 0;
        
        for ( int i = A.length - 2; i >= 0; i--)
        {
            for ( int j = 0; j < A[0].length; j++)
            {
                int sum = A[i+1][j];
                if (j+1 < A[0].length)
                    sum = Math.min(sum, A[i+1][j+1]);
                if ( j > 0)
                    sum = Math.min(sum, A[i+1][j-1]);
                
                A[i][j] = A[i][j] + sum;
            }
        }
        
        int minSum = Integer.MAX_VALUE;
        for (int n : A[0])
        {
            minSum = Math.min(minSum, n);
        }
        return minSum;
}
}

File: minimum_path_falling_sum.py
# Time Complexity: O(N^2), where N is the length of A.
# Space Complexity: O(1) in additional space complexity. 

# Yes, it ran on Leetcode successfully
# Problems: No

class Solution(object):
    def minFallingPathSum(self, A):

        row = len(A)
        col = row
        
        for r in range(1,row):
            for c in range(col):
                if c==0:
                    A[r][c] += min(A[r-1][c],A[r-1][c+1])
                elif c==col-1:
                    A[r][c] += min(A[r-1][c-1],A[r-1][c])
                else:
                    A[r][c] += min(A[r-1][c-1],A[r-1][c],A[r-1][c+1])
        return min(A[-1])